Rain Bird LNK WiFi Irrigation Implementation

jbarrancos opened the pull request:
This is a component which adds support for the Rain Bird LNK WiFi Irragation system.

## Description:
Adds component support for the Rain Bird Sprinkler / Irrigation system's LNK Wifi Module.
This component communicates directly towards the IP Address of the WiFi module it does not support the cloud. You can start/stop the irrigation and get the currenltly active zone.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3242

## Example entry for `configuration.yaml`:
```yaml

rainbird:
  host: '1.1.1.1'
  password: 'password'

```

jbarrancos committed:
Rainbird WiFi LNK Irrigation Implementation 

This is a component which adds support for the Rainbird WiFi LNK Irragation system.

homeassistant commented:
Hi @jbarrancos, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#9118).

Once you do that we will be able to review and accept this pull request.

Thanks!

jbarrancos committed:
Removed some whitelines

Removed some whitelines

jbarrancos committed:
Fixed some markup and removed not yet needed class

Fixed issues during hound run (markup) and a class which is obsolete

jbarrancos committed:
Markup fixes 

Markup fixes

jbarrancos committed:
PIP8 Fixes

PIP8 Fixes

jbarrancos committed:
Typo

jbarrancos committed:
pylint, coverage and requirement fix

pylint, coverage and requirement fix

jbarrancos committed:
pydocstyle fixes

pydocstyle fixes

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+
+I'm not a Python developer, so sorry for the bad code. I've developed it to
+control it from my domtica systems.
+"""
```
Please update the docstring to match the other platforms.

fabaff commented on homeassistant/rainbird.py:
```diff
@@ -0,0 +1,74 @@
+import homeassistant.helpers as helpers
```
Duplicate or miss-placed file.

jbarrancos replied:
Old

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+    Call from Home Assistant.
+
+    @param hass: default homeassistant hass class
+    @param config: default homeassistant config class
```
A simple docstring would enough for now.

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
```
Configuration variables needs to run through the validation:

https://home-assistant.io/developers/code_review_component/#configuration

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+    @param hass: default homeassistant hass class
+    @param config: default homeassistant config class
+    """
+    server = config[DOMAIN].get('stickip')
```
Please follow the common naming schema for configuration variables.

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+# Home Assistant Setup
+DOMAIN = 'rainbird'
+SERVER = ''
+PASSWORD = ''
```
Doesn't seem to be used.

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+
+REQUIREMENTS = ['pyrainbird==0.0.7']
+
+# Home Assistant Setup
```
No need for that comment.

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+    from pyrainbird import RainbirdController
+
+    controller = RainbirdController(_LOGGER)
+    controller.setConfig(server, password)
```
How do you ensure that the password is correct? What happens if the device is not accessible?

jbarrancos replied:
There are 2 failure scenario's. 0 is a NOT ACK from the controller, -1 is a failure (timeout, config etc after 1 retry). Debug level logging should give an indication towards the problem.

fabaff replied:
At the moment wouldn't the users end-up with a non-functional platform in their setup if their password is wrong? The users would only be able to recognize this problem after calling the service.

balloob replied:
Better to raise `PlatformNotReady` exception and setup will be re-tried after Home Assistant has launched (only available once migrated to a switch platform)

fabaff commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,105 @@
+        result = controller.stopIrrigation()
+        if result == 1:
+            _LOGGER.info("Stopped irrigation")
+            print("Success")
```
Stale print.

fabaff reviewed (changes requested):
Just some comments to get this review rolling.

jbarrancos committed:
Changed doc and added default schema

jbarrancos commented on homeassistant/rainbird.py:
```diff
@@ -0,0 +1,74 @@
+
+STATE_VAR = 'rainbird.activestation'
+
+def setup(hass, config):
```
old

jbarrancos commented on homeassistant/rainbird.py:
```diff
@@ -0,0 +1,74 @@
+    hass.services.register(DOMAIN, 'stop_irrigation', stopirrigation)
+
+    helpers.event.track_time_change(
+            hass, lambda _: hass.states.set(STATE_VAR, getirrigation()), 
```
old

jbarrancos commented on homeassistant/rainbird.py:
```diff
@@ -0,0 +1,74 @@
+
+    helpers.event.track_time_change(
+            hass, lambda _: hass.states.set(STATE_VAR, getirrigation()), 
+            year=None, month=None, day=None, hour=None, minute=None, second=[00,30]
```
old

arsaboo replied:
This is still not fixed. Double check.

jbarrancos commented:
Bit lost why travis is still using, what seems to be, an old version of the file. All local checks are OK. Flake, pydocs, etc.

arsaboo commented:
Are you sure? I still see the errors, for example, I see `year=None, month=None, day=None, hour=None, minute=None, second=[00,30]`

jbarrancos commented:
Yes. I've already commited it yesterday. Do I need to update the pull somehow to let it take the latest? 

 +    helpers.event.track_time_change(
 +        hass, lambda _: hass.states.set(STATE_VAR, getirrigation()),
 +        year=None, month=None, day=None,
 +        hour=None, minute=None, second=[00, 30]

arsaboo commented:
Push the updates to your repo so that the PR is updated.

jbarrancos commented:
It is stated in the list here above. The current "cce4a56" does not have any of the issues it is complaining about.

arsaboo commented:
You have `rainbird.py` in the wrong location. It should be under `components`

jbarrancos committed:
St*pid mistake.. wrong place

jbarrancos commented:
Thanks.. that was .. smart ;)

jbarrancos committed:
Recommit corrected

Recommit corrected

jbarrancos commented:
Hopefully now the checks will go well ;)

balloob commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,99 @@
+
+        return result
+    initialstatus = getirrigation()
+    hass.states.set(STATE_VAR, initialstatus)
```
Don't use `hass.states.set` directly, use entity instead.

balloob commented on homeassistant/components/rainbird.py:
```diff
@@ -0,0 +1,99 @@
+    hass.services.register(DOMAIN, 'start_irrigation', startirrigation)
+    hass.services.register(DOMAIN, 'stop_irrigation', stopirrigation)
+
+    helpers.event.track_time_change(
```
This is taken care for you if you implement it as an entity of the switch component.

balloob reviewed (changes requested):
This should be implemented as a platform for the switch component.

Instead of setting the state directly, extend the Switch Entity abstract base class.

frenck commented:
Little heads up, the PR for the documentation was riddled with wrongly spelling of the manufacturer and product name. Things I've seen a lot:
- `Rainbird` must be `Rain Bird`
- `WiFi LNK` -> `LNK WiFi`

The full product name should be: `Rain Bird LNK WiFi Module` according to the manufacturer.
I guess this mistakes might be showing up in the code base as well.

jbarrancos committed:
Changed component to use entity and switch

Changed component to use entity and switch.

jbarrancos committed:
Fixed issue with missing key

jbarrancos commented:
Thanks for the comments. I've tried to implement them to the best of my knowledge. @balloob @fabaff @arsaboo @frenck

jbarrancos committed:
Using latest module and fixed state issue

- pyrainbird 0.0.9 allows the override (if ever needed) connection retry/sleep
- Forces state towards the Entity when switching the switches. Gives better UI experience.

jbarrancos committed:
Hound issue "whitespace"

jbarrancos committed:
Updated requirements_all.txt

pvizeli commented:
Please move all code into `switch/rainbird.py` like `light/hue.py`

jbarrancos commented:
> "Please move all code into switch/rainbird.py like light/hue.py" 

The hue light is purely a "light". The Rain Bird controller, is a component which has multiple switches. I can migrate all code into the switch, but I will still need to put an Entity of the Controller inside the switch, so I don't have all switches polling towards the WiFi module.

If that isn't a problem, it's fine by me.

jbarrancos committed:
Moved all code into the switch component

Per request moved all the code inside the switch

pvizeli commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,224 @@
+        return self._state
+
+
+class RainbirdDevice(Entity):
```
Make no entity with this. Move that code into pyrainbird and make a better API interface for home-assistant and make simple switches. You can also store the API Object into `hass.data`. I don't know the rainbird but maybe you can read all available switch from interface and add this into hass without you need do that manual. After that you can use `customize` for change the friendly name.

jbarrancos replied:
Move it again ;) It took me way to long already to analyse the Rain Bird data and get an API of some sorts. I've already changed my HAAS project twice completely and I'm fine with it as it is. I can (as it now is) keep it in my custom_components. So take the pull or leave it.

jbarrancos replied:
I'm getting a bit lost on what to change now, as I keep changing it and still it seems to be going in the wrong direction. So same info on what I needed:

Rain Bird API allows to switch a valve on for a given duration.  So I need a switch + duration. There is no way to know which of the 4 - 8 -12 - 24 groups is connected. So that's why i wanted it configurable as most people won't have everything connected. 

Secondly I need to minimize the calls towards the API on the WiFi LNK module. When a call happens it takes about 10 seconds (sometimes longer) before a different connection can make requests, like the Rain Bird App. So as long as HAAS is the only one polling, it won't be an issue to have every switch polling. But most users will still want to use the app of Rain Bird itself and that is an issue. That's why I needed an additional layer/entity to handle requests and state.

For my own setup in HAAS I just use the Controller Entity and use customizations, scripts, automation scripts and custom sensors to interact with the 2 services exposed by this component. The switches are just there for people who want to have simple on / off sprinklers with a set duration in config.

So yes, I can make some changes, but I need to know exactly what it is you want.

jbarrancos commented:
Ping @pvizeli

jbarrancos committed:
Merge pull request #1 from home-assistant/dev

Update Fork

balloob commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,224 @@
+        _LOGGER.error("Error getting state. Possible configuration issues")
+        raise PlatformNotReady
+    else:
+        _LOGGER.info("Initialized Rain Bird Controller")
```
Please don't log this as info.

balloob commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,224 @@
+    """Set up Rain Bird switches over a Rain Bird controller."""
+    server = config.get(CONF_HOST)
+    password = config.get(CONF_PASSWORD)
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
```
You are not allowed to use the entity component. This is a platform for the switch entity component.

balloob commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,224 @@
+        @param call: should be a home assistant call object with data
+        station for Zone to sprinkle and duration for the time
+        """
+        _LOGGER.info("Requesting irrigation for " +
```
Please include less logging and limit logging to info to the really necessary. Note that service calls are already logged

balloob commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,224 @@
+        self._stations = {}
+
+        # For automation purposes add 2 services
+        def start_irrigation_call(call):
```
These should be moved inside the `turn_on` and `turn_off` methods of the entity above, making this class (which is not an entity btw) obsolete. You should not register services for this but instead have switch.turn_on and switch.turn_off do the work.

balloob reviewed (dismissed):
You have included parts of components and parts of a platform in a platform. All component specific things like usage of EntityComponent and registering services has to be removed. A platform should only include a small setup function and an Entity class definition that interacts with the 3rd party Python library

jbarrancos committed:
Limited to switch on/off

Limited to switch on/off
Lowered loglevel

jbarrancos committed:
Merge pull request #2 from jbarrancos/rainbird

Limited to switch on/off

jbarrancos committed:
Added dependency on pycrypto

Crypto Dependency missing found on virtualenv install. Added to dependecy rainbird.py

jbarrancos committed:
Requirements updated

jbarrancos committed:
Fixed dependency version

jbarrancos committed:
Split requirements per line

jbarrancos committed:
Dependency breaks build

jbarrancos committed:
Revert "Dependency breaks build"

This reverts commit c1aaed250ac686418f6af349694a7a3ce7a0ca5d.

jbarrancos committed:
Merge pull request #3 from home-assistant/dev

Get latest

jbarrancos committed:
Merge pull request #4 from jbarrancos/dev

Merge in latest

jbarrancos committed:
Removed requirement

jbarrancos committed:
typo

jbarrancos committed:
Requirements updated

jbarrancos committed:
Merge pull request #5 from jbarrancos/rainbird

Rainbird

jbarrancos committed:
Removed requirement

jbarrancos committed:
Merge pull request #6 from jbarrancos/rainbird

Removed requirement

fabaff commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,116 @@
+    devices = []
+    for dev_id, switch in config.get(CONF_SWITCHES).items():
+        devices.append(RainBirdSwitch(controller, switch, dev_id))
+    add_devices(devices)
```
Add ` True` -> https://home-assistant.io/developers/code_review_platform/#5-entity

fabaff commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,116 @@
+    for dev_id, switch in config.get(CONF_SWITCHES).items():
+        devices.append(RainBirdSwitch(controller, switch, dev_id))
+    add_devices(devices)
+    return True
```
Remove this.

fabaff commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,116 @@
+        self._devid = dev_id
+        self._zone = int(dev.get(CONF_ZONE))
+        self._name = dev.get(CONF_FRIENDLY_NAME, "Sprinker %s" % self._zone)
+        self._state = self.get_device_status()
```
Set it to `None` and let `update()` handle it.

fabaff commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,116 @@
+        self._rainbird = rb
+        self._devid = dev_id
+        self._zone = int(dev.get(CONF_ZONE))
+        self._name = dev.get(CONF_FRIENDLY_NAME, "Sprinker %s" % self._zone)
```
String formatting would be preferred.

fabaff commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,116 @@
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
```
This can be removed, as [polling](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py#L78) is the default.

fabaff commented on homeassistant/components/switch/rainbird.py:
```diff
@@ -0,0 +1,116 @@
+DOMAIN = 'rainbird'
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = vol.Schema({
```
Extend the schema -> https://home-assistant.io/developers/code_review_platform/#3-configuration

jbarrancos committed:
Fixed comments from @fabaff

Fxied issues raides

jbarrancos committed:
Fix req

jbarrancos committed:
lint error

jbarrancos committed:
lint ws

jbarrancos committed:
Merge pull request #7 from jbarrancos/rainbird

Fixed comments from @fabaff

pvizeli committed:
cleanup
